### DISCUSSION TITLE
NVIDIA DOCA East-West Overlay Encryption Reference Application - IPsec based strongSwan solution.

### DISCUSSION THREAD

**Discussion by benboreham** [strongswan/strongswan#964] [Q&A] [answered]
Hi all,

I'm having a bit of trouble with the strongswan example provided by Nvidia on this page: 
https://docs.nvidia.com/doca/sdk/east-west-overlay-encryption/index.html

I've attached my config files from the left and right DPU's

I'm able to initialize the strongswan connection without issue:

> root@localhost:/home/ubuntu# swanctl --load-all
> loaded ike secret 'ike-BF'
> no authorities found, 0 unloaded
> no pools found, 0 unloaded
> loaded connection 'BFL-BFR'
> successfully loaded 1 connections, 0 unloaded
> root@localhost:/home/ubuntu# swanctl -i --child bf
> [IKE] establishing CHILD_SA bf{6}
> [ENC] generating CREATE_CHILD_SA request 4 [ N(USE_TRANSP) SA No KE TSi TSr ]
> [NET] sending packet: from 3.3.3.2[500] to 3.3.3.3[500] (256 bytes)
> [NET] received packet: from 3.3.3.3[500] to 3.3.3.2[500] (256 bytes)
> [ENC] parsed CREATE_CHILD_SA response 4 [ N(USE_TRANSP) SA No KE TSi TSr ]
> [CFG] selected proposal: ESP:AES_GCM_16_128/CURVE_25519/EXT_SEQ
> [IKE] CHILD_SA bf{6} established with SPIs cfffdee5_i c824949b_o and TS 3.3.3.2/32[udp/4789] === 3.3.3.3/32[udp/4789]
> initiate completed successfully
> root@localhost:/home/ubuntu# 
> 

But when I send packets via netcat between the hosts I am still able to see the packets in plain text.

> Frame 34: 123 bytes on wire (984 bits), 123 bytes captured (984 bits)
> Linux cooked capture v1
> Internet Protocol Version 4, Src: 3.3.3.3, Dst: 3.3.3.2
> User Datagram Protocol, Src Port: 49396, Dst Port: 4789
> Virtual eXtensible Local Area Network
> Ethernet II, Src: Mellanox_e7:1d:da (0c:42:a1:e7:1d:da), Dst: Mellanox_e7:1e:b2 (0c:42:a1:e7:1e:b2)
> Internet Protocol Version 4, Src: 192.168.1.2, Dst: 192.168.1.1
> Transmission Control Protocol, Src Port: 55872, Dst Port: 5555, Seq: 1, Ack: 1, Len: 5
> Data (5 bytes)
> 
> 0000   00 00 00 01 00 06 0c 42 a1 e7 1d de 00 00 08 00   .......B........
> 0010   45 00 00 6b 5f 51 40 00 40 11 cf 26 03 03 03 03   E..k_Q@.@..&....
> 0020   03 03 03 02 c0 f4 12 b5 00 57 00 00 08 00 00 00   .........W......
> 0030   00 00 64 00 0c 42 a1 e7 1e b2 0c 42 a1 e7 1d da   ..d..B.....B....
> 0040   08 00 45 00 00 39 7f d7 40 00 40 06 37 94 c0 a8   ..E..9..@.@.7...
> 0050   01 02 c0 a8 01 01 da 40 15 b3 9c 2e c7 c2 25 9b   .......@......%.
> 0060   61 3e 80 18 01 f6 c8 00 00 00 01 01 08 0a ee 96   a>..............
> 0070   ce 86 3c 21 63 8e 74 65 73 74 0a                  ..<!c.**test**.

^^
Plaintext "test" is visible in the tcpdump

If I set up strongswan on the hosts (instead of on the DPUs) and remove the [udp/4789] setting in /etc/swanctl/conf.d/BFL.swanctl.conf and /etc/swanctl/conf.d/BFR.swanctl.conf the packets show as ESP and are encrypted in the tcpdump.

I suspect that I am missing some configuration but I'm not sure where to look next.

Attached config and ovs logs:
[strongswan udp issue.zip](https://github.com/strongswan/strongswan/files/8299440/strongswan.udp.issue.zip)

Any assistance you can provide would be greatly appreciated.

**Comment by tobiasbrunner** [comment, marked as answer]
As you already noticed, the packets don't match the traffic selectors (source port 49396 doesn't match the selector for 4789), so they are not encrypted. How is the source port for VXLAN determined? Is that configured somewhere? Is there some NAT involved in your case?

**Reply by benboreham** [reply]
Thanks for your prompt reply tobiasbrunner

I was under the impression that the OVS config was handling the VXLAN with this setting:

> root@localhost:/home/ubuntu# ovs-vsctl show
> ...
> ...
> ...
>         Port vxlan11
>             Interface vxlan11
>                 type: vxlan
>                 options: {dst_port="4789", key="100", local_ip="3.3.3.2", remote_ip="3.3.3.3"}
> ... 
> ...
> ...               

There is no NAT in my setup, the two DPUs are connected back to back.

**Reply by tobiasbrunner** [reply]
That's only the destination port. And as far as I can tell, the kernel uses dynamic source ports for VXLAN based on a hash of the tunneled packet (within a range, which is configurable, however, it can't be set to a single port as that will cause a fallback to the default range for ephemeral source ports, see `udp_flow_src_port()`).

So it seems the strongSwan config NVIDIA shows in their example can't actually work. In fact, if you look at the log output there, the negotiated CHILD_SA has no protocol/port selectors (`CHILD_SA bf{1} established with SPIs ce543905_i c60e98a2_o and TS 192.168.50.1/32 === 192.168.50.2/32`). They might have added those selectors to the config later without actually testing them (their config also has `/24` traffic selectors, which is not reflected in the log output and doesn't make sense for host-to-host connections with transport mode, for those, using `local|remote_ts = dynamic[...]` is the better approach as that will automatically get replaced with the `/32` IP addresses of the hosts).

Anyway, to handle this properly, configure two child configs (e.g. `bf-out` and `bf-in`), one with `local_ts = ...[udp]` and `remote_ts = ...[udp/4789]` and the other with these selectors reversed (i.e. `local_ts = ...[udp/4789]` and `remote_ts = ...[udp]`). It can't be handled with a single CHILD_SA because configuring both traffic selectors (i.e. `local_ts = ...[udp], ...[udp/4789]` and `remote_ts = ...[udp/4789], ...[udp]`) would get them narrowed to `...[udp] === ...[udp]`, i.e. all UDP traffic would get tunneled. An alternative approach would be to remove the protocol/port selectors (for host-to-host connection you could actually remove `local|remote_ts`  completely as they default to the IP addresses of the hosts) and instead configure marks and then use firewall rules to selectively tunnel only VXLAN packets (e.g. based on their destination port).
